src/psyclone/dynamo0p3.py

```diff
@@ -6156,6 +6157,28 @@ def __init__(self):
         self._cma_operation = None
         self._is_intergrid = False  # Whether this is an inter-grid kernel
 
+    def reference_accesses(self, var_accesses):
+        '''Get all variable access information. All accesses are marked
+        according to the kernel metadata
+
+        :param var_accesses: VariablesAccessInfo instance that stores the\
+            information about variable accesses.
+        :type var_accesses: \
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+        '''
+        for arg in self.arguments.args:
```

Comment:
I started to wonder whether we couldn't use the same `reference_accesses` routine for both gocean1.0 and dynamo0.3 and then I realised that you're using the `vector_size` property to decide whether a kernel argument is an array. I suspect this isn't actually what you want to do since the naming is a bit confusing. In Kernel meta-data a user may specify that a kernel takes a 'vector' of fields by doing, e.g. "GH_FIELD*3". This translates into the kernel having 3 field arguments, each of which is then an array. I think you should really check on the `type` property of the argument - if it is anything other than a scalar (`gh_real` or `gh_integer` in dynamo0p3) then you have an array.

Reply:
Let me retell the odyssey here ... sit back, grab some popcorn 🍿 and relax ...

I liked the idea of unifying this ... so I added a special `is_scalar()` function (since scalars have a different `type` depending on API), which could then be used API independent. 

Then I discovered that in GOcean there are two different classes for arguments (GOKernelArgument and GOKernelGridArgument), and I didn't have `is_scalar()` for the latter. So I extended the data structure GRID_PROPERTY_DICT to included scalar/array information and added the third `is_scalar()`. While doing this and changing my test case to include a grid property, it turns out that (at least) the opencl code creation is incorrect if e.g. go_grid_x_min_index is used (the grid_property_dict maps this to "subdomain%internal%xstart" ... I assume Andy can see where this goes wrong then 😢 ). So I fixed that, too, and updated the test cases.
Next, I then looked at the variable access information, which was also not correct (using e.g. `xstart` as read variable in the example above, which makes no sense at all), so  I handled grid properties separately to instead mark `<some_field>%data%subdomain%internal%xstart` as READ (I'll leave the user defined type handling for latter - you can obviously argue that we should mark field, field%data, ... all as READ??).

So as a result of all this refactoring, the two implementations are still different and not in the base class, since I can't see an easy way of achieving this due to the special handling for grid properties (unless we do a lot more refactoring of the Argument classes, which imho should be a separate pull request).

Am happy to discuss this 😅

Reply:
Thanks for explaining (in a comment elsewhere).